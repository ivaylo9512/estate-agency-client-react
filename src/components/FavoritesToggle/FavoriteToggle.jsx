import { memo, useEffect, useState } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { getPropertyFavoriteState, addFavorite, removeFavorite } from 'app/slices/toggleFavorite';
import LoadingIndicator from '../LoadingIndicator/LoadingIndicator';

const FavoriteToggle = memo(({property}) => {
    const [isFavorite, setIsFavorite] = useState(property.isFavorite);
    const favoriteState = useSelector(getPropertyFavoriteState(property.id));
    const dispatch = useDispatch();

    const toggleFavorite = (property) => {
        if(!isFavorite){
            return dispatch(addFavorite(property.id));
        }
        dispatch(removeFavorite(property.id));
    }

    useEffect(() => {
        if(!favoriteState || favoriteState.isLoading){
            return;
        }

        setIsFavorite(favoriteState.isFavorite)
    },[favoriteState])

    return(
        <button onClick={() => toggleFavorite(property)}>{ 
            favoriteState && favoriteState.isLoading 
                ? <LoadingIndicator />
                : isFavorite 
                    ? 'remove'
                    : 'add'
        }
        </button>
    )

})
export default FavoriteToggle