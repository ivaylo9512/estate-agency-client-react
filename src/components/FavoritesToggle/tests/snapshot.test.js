import FavoriteToggle from '../FavoriteToggle';
import * as Redux from 'react-redux';
import { shallow } from 'enzyme';

describe('FavoriteToggle snapshot tests', () => {
    let selectorSpy;

    beforeAll(() => {
        jest.spyOn(Redux, 'useDispatch').mockReturnValue(jest.fn());

        selectorSpy = jest.spyOn(Redux, 'useSelector');
    })

    const createWrapper = (state) => {
        selectorSpy.mockReturnValue(state.favoritesState);

        return shallow(
            <FavoriteToggle property={ state.property }/>
        )
    }

    it('should match snapshot with favoritesState and isLoading to true', () => {
        const wrapper = createWrapper({ favoritesState: { isLoading: true }, property: { isFavorite: true }});

        expect(wrapper).toMatchSnapshot();
    })

    it('should match snapshot with favoritesState and isLoading to false and isFavorite to true', () => {
        const wrapper = createWrapper({ favoritesState: { isLoading: false }, property: { isFavorite: true }});

        expect(wrapper).toMatchSnapshot();
    })

    it('should match snapshot with favoritesState and isLoading to false and and isFavorite to false', () => {
        const wrapper = createWrapper({ favoritesState: { isLoading: false }, property: { isFavorite: false }});

        expect(wrapper).toMatchSnapshot();
    })

    it('should match snapshot with undefined favoritesState', () => {
        const wrapper = createWrapper({ property: { isFavorite: true }});

        expect(wrapper).toMatchSnapshot();
    })
})