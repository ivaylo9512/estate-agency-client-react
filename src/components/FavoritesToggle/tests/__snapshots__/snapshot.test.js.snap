// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FavoriteToggle snapshot tests should match snapshot with favoritesState and isLoading to false and and isFavorite to false 1`] = `
<button
  onClick={[Function]}
>
  add
</button>
`;

exports[`FavoriteToggle snapshot tests should match snapshot with favoritesState and isLoading to false and isFavorite to true 1`] = `
<button
  onClick={[Function]}
>
  remove
</button>
`;

exports[`FavoriteToggle snapshot tests should match snapshot with favoritesState and isLoading to true 1`] = `
<button
  onClick={[Function]}
>
  <LoadingIndicator />
</button>
`;

exports[`FavoriteToggle snapshot tests should match snapshot with undefined favoritesState 1`] = `
<button
  onClick={[Function]}
>
  remove
</button>
`;
